### DISCUSSION TITLE
Give the ability to configure the url structure for locales.

### DISCUSSION THREAD

**Discussion by SimplyComplexable** [vercel/next.js#46662] [Ideas]
### Goals

1. Give more flexibility in the url structure for locales without the need for middleware and/or other more complicated solutions.

### Non-Goals

1. Any use-cases for more complex locale structures beyond a simple basePath prefix.

### Background

Currently, locales and the URL paths that represent those locales are tightly coupled. If you want to enable a `en-US` locale, using the out-of-the-box solution, any routes for that specific locale will be prepended with `/en-us`. However, in some cases, websites need to use a URL structure that is distinct from the actual locale. A good example is the Apple website which uses the pattern `/[region]/[language]`. This is the Canadian French site https://www.apple.com/ca/fr/.  This is actually exactly the situation I'm trying to solve for, but for the Traeger website, https://www.traeger.com/ca/fr.

On the Traeger website we're solving for this using middleware that creates rewrites for all the different locales. This works, but it causes some other issues and just creates maintenance issues trying to reconcile, back and forth, between the two. Ideally, this pattern could just be directly configured within Next and remove any need for more brittle solutions.

### Proposal

I think a pretty straightforward option would be to provide an alternative API that uses objects rather than strings for the locales configuration in the next.config. 
```
i18n: {
  locales: [
    {
      locale: 'en-US',
      path: 'us/en'
    },
    {
      locale: 'fr-CA',
      path: 'ca/fr'
    },
    {
      locale: 'fr',
      path: 'fr'
    }
  ]
}
```
If the path and locale should be the same, the existing config (with just strings instead of objects) could still be used, allowing for backwards compatibility. Otherwise the actual local would be the `locale` property within the object and the path would be the `path` property.

**Comment by noidar** [comment]
Hey @SimplyComplexable, I am interested how you solved the problem with the country and name being the same?
I have the same situation I am using next-i18next like sitename.com/de/de sitename.com/de/at
ex: https://github.com/noidar/next-18next-router-problem
but I have to fix with some helper functions when it comes to the same location an language.
Thanks.

**Reply by SimplyComplexable** [reply]
I'm not sure I'm following correctly, but we're using middleware that does rewrites for all the locales. This is roughly the code that does it.
```
    if (!location.isDefault) {
      url.pathname = `${location.locale}${pathname.replace(location.geoLoc, '')}`;
      url.locale = location.locale;
      return NextResponse.rewrite(url);
    }
```

Where location looks roughly like one of these.
```
  {
    "geoLoc": "/fr/fr",
    "country": "/fr",
    "locale": "fr-FR",
    "generalPage": "general-fr",
    "productGeo": "FR",
    "name": "France",
  }

  {
    "geoLoc": "/ca/en",
    "country": "/ca",
    "locale": "en-CA",
    "productGeo": "CA",
    "name": "Canada",
  }
```
Hopefully that helps!

**Reply by SimplyComplexable** [reply]
Oh also, I just wanted to mention that you can use a [matcher](https://nextjs.org/docs/advanced-features/middleware#matcher) on middleware which routes the middleware will run on (and save money on those edge functions), and remove the need to do the checks  inside of the middleware itself.

**Comment by silas-v1** [comment]
Good proposal. I would also consider this as an issue.
